AI/BattleAI/BattleExchangeVariant.cpp

```diff
@@ -0,0 +1,682 @@
+	}
+
+	if(defender->alive() && defender->ableToRetaliate() && !counterAttacksBlocked && !shooting)
+	{
+		if(retalitation.second != 0)
+		{
+			auto retalitationDamage = (retalitation.first + retalitation.second) / 2;
+			attackerDpsReduce = AttackPossibility::calculateDpsReduce(defender.get(), attacker.get(), retalitationDamage, cb);
+
+			if(!evaluateOnly)
+			{
+#if BATTLE_TRACE_LEVEL==1
+				logAi->trace(
+					"%s -> %s, retalitation, dps: %d, %d",
+					defender->getDescription(),
+					attacker->getDescription(),
+					retalitationDamage,
+					attackerDpsReduce);
+#endif
+
+				if(isOurAttack)
+				{
+					dpsScore -= attackerDpsReduce;
+					attackerValue[attacker->unitId()].isRetalitated = true;
+				}
+				else
+				{
+					dpsScore += attackerDpsReduce;
+					attackerValue[defender->unitId()].value += attackerDpsReduce;
+				}
+
+				attacker->damage(retalitationDamage);
+				defender->afterAttack(false, true);
+			}
+		}
+	}
+
+	auto score = defenderDpsReduce - attackerDpsReduce;
+
+#if BATTLE_TRACE_LEVEL==1
+	if(!score)
+	{
+		logAi->trace("Zero %d %d", defenderDpsReduce, attackerDpsReduce);
+	}
+#endif
+
+	return score;
+}
+
+EvaluationResult BattleExchangeEvaluator::findBestTarget(const battle::Unit * activeStack, PotentialTargets & targets, HypotheticBattle & hb)
+{
+	EvaluationResult result(targets.bestAction());
+
+	updateReachabilityMap(hb);
+
+	for(auto & ap : targets.possibleAttacks)
+	{
+		int64_t score = calculateExchange(ap, targets, hb);
+
+		if(score > result.score)
+		{
+			result.score = score;
+			result.bestAttack = ap;
+		}
+	}
+
+	if(!activeStack->waited())
+	{
+#if BATTLE_TRACE_LEVEL==1
+		logAi->trace("Evaluating waited attack for %s", activeStack->getDescription());
+#endif
+
+		hb.getForUpdate(activeStack->unitId())->waiting = true;
+		hb.getForUpdate(activeStack->unitId())->waitedThisTurn = true;
+
+		updateReachabilityMap(hb);
+
+		for(auto & ap : targets.possibleAttacks)
+		{
+			int64_t score = calculateExchange(ap, targets, hb);
+
+			if(score > result.score)
+			{
+				result.score = score;
+				result.bestAttack = ap;
+				result.wait = true;
+			}
+		}
+	}
+
+	return result;
+}
+
+MoveTarget BattleExchangeEvaluator::findMoveTowardsUnreachable(const battle::Unit * activeStack, PotentialTargets & targets, HypotheticBattle & hb)
+{
+	MoveTarget result;
+	BattleExchangeVariant ev;
+
+	if(targets.unreachableEnemies.empty())
+		return result;
+
+	updateReachabilityMap(hb);
+
+	auto dists = cb->getReachability(activeStack);
+	auto speed = activeStack->Speed();
+
+	for(const battle::Unit * enemy : targets.unreachableEnemies)
+	{
+		int64_t stackScore = EvaluationResult::INEFFECTIVE_SCORE;
+
+		std::vector<const battle::Unit *> adjacentStacks = getAdjacentUnits(enemy);
+		auto closestStack = *vstd::minElementByFun(adjacentStacks, [&](const battle::Unit * u) -> int64_t
+			{
+				return dists.distToNearestNeighbour(activeStack, u) * 100000 - activeStack->getTotalHealth();
```

Comment:
instead of this, does it make to evaluate all score for all enemies with the same distance?
You can process all stacks with certain distance and compute score with calculateExchange and then select the best choice